Refactor: New KonfluxClient class for calling Konflux APIs

vfreex opened the pull request:
- Move code for Konflux interactions into separate class `KonfluxClient`.
- Use LRU cache for certain APIs.

vfreex commented:
@ashwindasr Thanks for the review. I've pushed another commit to fix the flake8 errors. Test build: https://art-jenkins.apps.prod-stable-spoke1-dc-iad2.itup.redhat.com/job/aos-cd-builds/job/build%252Focp4-konflux/362/console

locriandev commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+    async def _patch(self, manifest: dict):
+        """ Patch a resource.
+
+        :param dyn_client: The dynamic client.
```
param `dyn_client` is not defined

locriandev commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+    async def _replace(self, manifest: dict):
+        """ Replace a resource.
+
+        :param dyn_client: The dynamic client.
```
param `dyn_client` is not defined

vfreex replied:
done

locriandev commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+    async def _delete(self, api_version: str, kind: str, name: str, namespace: str):
+        """ Delete a resource.
+
+        :param dyn_client: The dynamic client.
```
param `dyn_client` is not defined

locriandev commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+    async def _create_or_patch(self, manifest: dict):
+        """ Create or patch a resource.
+
+        :param dyn_client: The dynamic client.
```
param dyn_client is not defined

locriandev commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+    async def _create_or_replace(self, manifest: dict):
+        """ Create or replace a resource.
+
+        :param dyn_client: The dynamic client.
```
param dyn_client is not defined

locriandev commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+        output_image: str,
+        building_arches: Sequence[str],
+        git_auth_secret: str = "pipelines-as-code-secret",
+        additional_tags: Sequence[str] = [],
```
default value for `additional_tags` sis mutable. In Python, mutable args are evaluated only once at function definition time, so calling the function with a different value of the argument will affect all subsequent calls of that function. Better use `None` instead

vfreex replied:
Done

locriandev commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+        """
+        Wait for a PipelineRun to complete.
+
+        :param dyn_client: The dynamic client.
```
param dyn_client is not defined

locriandev commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+    @staticmethod
+    def _new_component(name: str, application: str, component_name: str,
+                       image_repo: Optional[str], source_url: Optional[str], revision: Optional[str]) -> dict:
+        obj = {
```
Consider storing this definition as a jinja template, instead of hardcoding it. This could make it easier to find and maintain

locriandev commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+
+        # https://konflux.pages.redhat.com/docs/users/how-tos/configuring/overriding-compute-resources.html
+        # ose-installer-artifacts fails with OOM with default values, hence bumping memory limit
+        obj["spec"]["taskRunSpecs"] = [{
```
Same as previous comment, I think storing these definitions in template files would be best. Also ok with a follow up PR

vfreex commented on doozer/doozerlib/backend/konflux_client.py:
```diff
@@ -0,0 +1,507 @@
+                    if e.status == 410:
+                        # If the last result is too old, an `ApiException` exception will be thrown with
+                        # `code` 410. In that case we have to recover by retrying without resource_version.
+                        self._logger.debug("%s: Resource version %s is too old. Recovering...", pipelinerun_name, resource_version)
```
Line 476 is called without a resource version, so it should never raises a 410 APIException. But for robustness, I've removed the `resource_version` reference from logging.

vfreex commented:
@locriandev Thanks for the review. I've addressed your review suggestions again except the one for moving application/component definition to jinja files, which can be done in separate PRs.

vfreex commented:
Another test run https://art-jenkins.apps.prod-stable-spoke1-dc-iad2.itup.redhat.com/job/aos-cd-builds/job/build%252Focp4-konflux/439/console

locriandev reviewed (commented):
/lgtm
